### ISSUE TITLE
On protocol & implementation specifications

### ISSUE THREAD

**Issue by cwgoes** [cosmos/ibc#23]
I think we should separate *protocol* specifications from *implementation* specifications.

What makes a spec one or the other?

Protocol specs outline the abstract logic required to implement a protocol, including algorithms, asymptotic complexity, protocol-layer invariants, guarantees, or requirements, and pseudocode descriptions of logic. Protocol specifications should be sufficiently precise for independent implementations of the specification to be compatible.

Example protocol specs:
- Algorithm pseudocode in the [Tendermint paper](https://arxiv.org/pdf/1807.04938.pdf) (page 6)
- The [signed message spec](https://github.com/cosmos/cosmos-sdk/blob/develop/docs/spec/_ics/ics-030-signed-messages.md) (except for one Golang struct definition, which should be removed)
- [Tendermint versioning spec](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-016-protocol-versions.md)

Implementation specs describe the particular data structures and functions in a specific language which will be used to implement a previously defined protocol. They may include varying amounts of actual code, but generally concern themselves with logical clarity in the intended implementation language, proper abstraction design, practical efficiency considerations, etc.

Example implementation specs:
- [SDK x/bank spec](https://github.com/cosmos/cosmos-sdk/tree/develop/docs/spec/bank)
- [SDK x/gov spec](https://github.com/cosmos/cosmos-sdk/tree/develop/docs/spec/governance)

I think we should either reserve the ICS repo for protocol specs, and keep implementation specs in the implementation repositories (e.g. the Cosmos SDK), or at minimum keep implementation specs in a separate folder, so future users interested in implementing the protocol can find language-agnostic high-level documentation.

**Comment by cwgoes** [comment]
I covered some of this in [ICS 1](https://github.com/cosmos/ics/pull/17) under "What is an ICS?".

**Comment by mossid** [comment]
I think we can store one protocol specification and one (possibly multiple) implementation specification(s) in a `ics-XXX-*/` together. For example, `ics-003-connection-semantics/{protocol.md, sdk-implementation.md, rust-implementation.md, etc.}`.

**Comment by cwgoes** [comment]
We could, but in the future when there are many implementations of a particular protocol (IBC connection version negotiation, for example) I think that might get messy - better to leave the implementation specifications in their own repositories (e.g. the Cosmos SDK repo) and link to them in the protocol spec (there is a sub-section for implementations in [ICS 1](https://github.com/cosmos/ics/pull/17/files#diff-66b76f0e4ccd9548753d28da72f501eeR97)).

**Comment by rigelrozanski** [comment]
from slack 
> we should distinguish between implementation spec and state-machine-expected-behaviour specs - from my experience, implementation evolves naturally as the code matures and we should not require any more process than we currently have - although we should maintain an implementation spec as well. When developing the code, so long as the expected behaviour stays the same (or is corrected to expected behaviour) as the existing implementation (or approved ICS) these implementations should not require going through the ICS proposal process

**Comment by cwgoes** [comment]
> we should distinguish between implementation spec and state-machine-expected-behaviour specs - from my experience, implementation evolves naturally as the code matures and we should not require any more process than we currently have - although we should maintain an implementation spec as well. When developing the code, so long as the expected behaviour stays the same (or is corrected to expected behaviour) as the existing implementation (or approved ICS) these implementations should not require going through the ICS proposal process

Agreed. Implementation specs are out-of-scope of the ICS process & this repository - those are left to the implementers (e.g. Cosmos SDK) to choose and execute upon.

**Comment by ethanfrey** [comment]
Yes, I like the focus on protocol specs

**Comment by cwgoes** [comment]
Also see https://github.com/cosmos/ics/issues/47 for several historical protocol specifications for inspiration.

**Comment by Liamsi** [comment]
I' also all in favour of not including implementation specs in this repo. If anything we could recommend implementers to include a design document / implementation spec in their repo.

**Comment by mossid** [comment]
We need a "concrete spec"s which are the implementations of `abstract spec`s, for example, ICS23 Accumulator is an abstract spec where Merkle tree is a concrete spec which satisfies ICS23, ICS2 Consensus Verification is an abstract spec where Tendermint Lightclient is a concrete spec which satisfies ICS2.

Separated from the actual implementations, it seems that storing concrete specs under the abstract specs looks organized:

```
ICS/
- ICS2: Consensus Verification
  - ICS2-1: Tendermint Lightclient(link to tendermint/lite, etc. under "implementations")
  - ICS2-2: Finality Pegzone(link to finality pegzone implementations)
- ICS23: Accumulator
  - ICS23-1: Merkle tree(link to iavl, merkle patricia, etc. under "implementations")
  - ICS23-2: Another Accumulator(link to their implementations...)
```
clearer then

```
- ICS2: Consensus Verification
- ICS7: Tendermint Lightclient
- ICS22: Finality Pegzone
- ICS23: Accumulator
- ICS??: Merkle Tree
``

which makes the relationship less noticeable

The distinction between concrete specs and implementations specs are little blurry, and I think that the implementation specs should live with their code, but the concrete specs will be defined as ICSs any way, better structure them in some sort of hierarchical way

**Comment by cwgoes** [comment]
> The distinction between concrete specs and implementations specs are little blurry, and I think that the implementation specs should live with their code, but the concrete specs will be defined as ICSs any way, better structure them in some sort of hierarchical way

This is an excellent point - maybe we should call them "interface" specifications and "concrete" specifications (both are protocols, not implementations, since we're not including actual code)? 

I agree with the hierarchical folder structure.

We should look at the current ICS set and figure out which will be "concrete" specs, as you've started to do.

**Comment by cwgoes** [comment]
Note from @jaekwon - call it "ICS 2: Tendermint"

**Comment by cwgoes** [comment]
This has been sufficiently clarified, I think.

**Comment by gjermundgaraba** [comment]
This issue was mistakenly reopened during some integration work, so closing again now.
